src/cargo/ops/cargo_fetch.rs

```diff
@@ -57,6 +58,19 @@ pub fn fetch<'a>(
             .map(|(id, _deps)| id);
         deps_to_fetch.extend(deps);
     }
+
+    // If -Zbuild-std was passed, download dependencies for the standard library.
+    if let Some(crates) = &config.cli_unstable().build_std {
+        // Fetch the dependencies for `test` just in case
+        // (cargo could later decide to build libtest even if it wasn't specified in the flags)
+        let mut crates = crates.clone();
+        crates.push("test".to_string());
```

Comment:
If this is being explicitly added here then I think that this should mirror something that happens in `cargo_compile`. IIRC `test` or `proc_macro` is explicitly added there, and whatever does that should be used to add the name `test` here. We want to avoid hardcoding crate names in multiple places.

Reply:
The problem is that `cargo_compile` is calculating this based off the command you passed, and `fetch` doesn't know ahead of time which unit you want to build. So it can't use the same logic.https://github.com/rust-lang/cargo/blob/7df467c29103db877c8a27b60074dbe5dbbc0799/src/cargo/ops/cargo_compile.rs#L541-L553

Reply:
I think that all that logic can be shard, and the `units.iter().any(...)` can be replaced with a boolean taken as an argument which is `true` for `cargo fetch` and calculated as-is for `cargo build`

Reply:
I'm still not exactly sure what you're looking for, but I pushed a commit that only hard-codes the crate name in a single function in `standard_lib`: https://github.com/rust-lang/cargo/pull/10129/commits/036146bc4070ce165a11df77fedd2e40756ed1bb